From 62da149e7dfdc68e4896e7b8de656af5291daaf8 Mon Sep 17 00:00:00 2001
From: Guojing Yuan <guojing@google.com>
Date: Tue, 8 Jul 2025 18:56:34 +0000
Subject: [PATCH 07/55] [CDM] Limit NLS revoke to applicable roles.

Bug: 421919075
Flag: EXEMPT bugfix
Test: CTS
(cherry picked from commit 8ccb7103a2d58ec2cee0a1cf5f8c7fefa5e8e8ee)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:ffa1ea81ac0c11f9d0c5a81b6e85772c23743210
Merged-In: Icd3cdb3234a516ff7f752f39d7b1d8e6f5cde1ef
Change-Id: Icd3cdb3234a516ff7f752f39d7b1d8e6f5cde1ef
---
 .../association/DisassociationProcessor.java  | 22 +++++++++++--------
 .../server/companion/utils/RolesUtils.java    |  6 +++++
 2 files changed, 19 insertions(+), 9 deletions(-)

diff --git a/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java b/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
index 928a08afb..81bfe5fc6 100644
--- a/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
+++ b/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
@@ -20,6 +20,7 @@ import static android.app.ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREG
 import static android.companion.AssociationRequest.DEVICE_PROFILE_AUTOMOTIVE_PROJECTION;
 
 import static com.android.internal.util.CollectionUtils.any;
+import static com.android.server.companion.utils.RolesUtils.NLS_PROFILES;
 import static com.android.server.companion.utils.RolesUtils.removeRoleHolderForAssociation;
 
 import static java.util.concurrent.TimeUnit.DAYS;
@@ -147,15 +148,18 @@ public class DisassociationProcessor {
         // Revoke NLS if the last association has been removed for the package
         Binder.withCleanCallingIdentity(() -> {
             if (mAssociationStore.getAssociationsByPackage(userId, packageName).isEmpty()) {
-                Intent nlsIntent = new Intent(
-                        NotificationListenerService.SERVICE_INTERFACE);
-                List<ResolveInfo> matchedServiceList = mContext.getPackageManager()
-                        .queryIntentServicesAsUser(nlsIntent, /* flags */ 0, userId);
-                for (ResolveInfo service : matchedServiceList) {
-                    if (service.getComponentInfo().getComponentName().getPackageName()
-                            .equals(packageName)) {
-                        mNotificationManager.setNotificationListenerAccessGranted(
-                                service.getComponentInfo().getComponentName(), false);
+                if (association.getDeviceProfile() != null
+                        && NLS_PROFILES.contains(association.getDeviceProfile())) {
+                    Intent nlsIntent = new Intent(
+                            NotificationListenerService.SERVICE_INTERFACE);
+                    List<ResolveInfo> matchedServiceList = mContext.getPackageManager()
+                            .queryIntentServicesAsUser(nlsIntent, /* flags */ 0, userId);
+                    for (ResolveInfo service : matchedServiceList) {
+                        if (service.getComponentInfo().getComponentName().getPackageName()
+                                .equals(packageName)) {
+                            mNotificationManager.setNotificationListenerAccessGranted(
+                                    service.getComponentInfo().getComponentName(), false, false);
+                        }
                     }
                 }
             }
diff --git a/services/companion/java/com/android/server/companion/utils/RolesUtils.java b/services/companion/java/com/android/server/companion/utils/RolesUtils.java
index c63577ef8..9a582602e 100644
--- a/services/companion/java/com/android/server/companion/utils/RolesUtils.java
+++ b/services/companion/java/com/android/server/companion/utils/RolesUtils.java
@@ -90,6 +90,12 @@ public final class RolesUtils {
             DEVICE_PROFILE_VIRTUAL_DEVICE, List.of(PERMISSION_CREATE_VIRTUAL_DEVICE,
                             PERMISSION_NEARBY_DEVICES, PERMISSION_POST_NOTIFICATIONS)
     );
+    
+    public static final Set<String> NLS_PROFILES = Set.of(
+            DEVICE_PROFILE_WATCH,
+            DEVICE_PROFILE_GLASSES,
+            DEVICE_PROFILE_APP_STREAMING,
+            DEVICE_PROFILE_COMPUTER);
 
     private static final Set<String> ROLELESS_DEVICE_PROFILES;
     static {
-- 
2.52.0

